### PR TITLE
Added feature required to entity relationship

### LINKED ISSUE #3417: [FEATURE] Possibility of add "required" in a one to many relationship
Currently if I create a one to many relationship (Owner-Car), I can create records in the many part with the relationship field empty (i.e. Car without an owner).

I was told [in the jhipster domain language issue tracker](https://github.com/jhipster/jhipster-domain-language/issues/8) that if Jhipster implements this, JHipster UML and the JDL would incorporate that. 
Any chance it is considered?

 **JHipster Version(s)** 
3.0

### PR DESCRIPTION
**Pull request by dornala**
Change Log:
-Added additional question's asking user for more specific details.
-Added "Required" when Entity is choosen with relationship Many to one.
-Add required in client field (To save an value of entity only when relationship is selected)

fix #3417

### CODE REVIEW

**Comment by dornala** [comment]
Wow that's bad..:(.

File: generators/entity/index.js [hunk 1]
```diff
@@ -273,7 +273,14 @@ module.exports = EntityGenerator.extend({
         if (this.relationships.length > 0) {
             this.log(chalk.white('Relationships'));
             this.relationships.forEach(function (relationship) {
-                this.log(chalk.red(relationship.relationshipName) + ' ' + chalk.white('(' + _.upperFirst(relationship.otherEntityName) + ')') + ' ' + chalk.cyan(relationship.relationshipType));
+            	var validationDetails = '';
```
Comment (deepu105) [inline generators/entity/index.js]: indent is wrong

File: generators/entity/index.js [hunk 2]
```diff
@@ -273,7 +273,14 @@ module.exports = EntityGenerator.extend({
...
+                var relationshipValidate = _.isArray(relationship.relationshipValidateRules) && relationship.relationshipValidateRules.length >= 1;
```
Comment (deepu105) [inline generators/entity/index.js]: use simple checks, the below is sufficient, no need of `relationshipValidate ` and stuff
```
if (relationship.relationshipValidateRules && relationship.relationshipValidateRules.indexOf('required') !== -1) {

}
```

File: generators/entity/index.js [hunk 3]
```diff
@@ -563,11 +570,11 @@ module.exports = EntityGenerator.extend({
                         response.fieldType === 'BigDecimal' ||
                         response.fieldTypeBlobContent === 'text');
                 },
-                type: 'checkbox',
-                name: 'fieldValidateRules',
-                message: 'Which validation rules do you want to add?',
-                choices: [
-                    {
+                    type: 'checkbox',
+            name: 'fieldValidateRules',
```
Comment (deepu105) [inline generators/entity/index.js]: indent changed, this code shouldnt be changed for this PR

File: generators/entity/index.js [hunk 4]
```diff
@@ -925,6 +932,33 @@ module.exports = EntityGenerator.extend({
             },
             {
                 when: function (response) {
+                    return response.relationshipAdd === true &&
+                        response.relationshipType === 'many-to-one';
```
Comment (deepu105) [inline generators/entity/index.js]: required could be set for one-to-one as well rite? and technically I can even have required for other types of relationships as well

File: generators/entity/index.js [hunk 5]
```diff
@@ -1566,6 +1601,12 @@ module.exports = EntityGenerator.extend({
                     this.fieldsContainManyToOne = true;
                 }
 
+                if (_.isArray(relationship.relationshipValidateRules) && relationship.relationshipValidateRules.length >= 1) {
```
Comment (deepu105) [inline generators/entity/index.js]: just do `if (relationship.relationshipValidateRules)` thats enough, it will give false if `relationship.relationshipValidateRules` is undefined, null or empty

File: generators/entity/index.js [hunk 6]
```diff
@@ -1743,7 +1784,7 @@ module.exports = EntityGenerator.extend({
     install: function () {
         var injectJsFilesToIndex = function () {
             this.log('\n' + chalk.bold.green('Running gulp Inject to add javascript to index\n'));
-            this.spawnCommand('gulp', ['inject:app']);
+            this.spawnCommand('gulp', ['inject']);
```
Comment (deepu105) [inline generators/entity/index.js]: why is this changed? this shouldnt be changed

File: generators/entity/templates/src/main/java/package/domain/_Entity.java [hunk 7]
```diff
@@ -109,13 +109,15 @@
     <%_ }
     }
     for (idx in relationships) {
-        var otherEntityRelationshipName = relationships[idx].otherEntityRelationshipName,
+		var relationshipValidate = relationships[idx].relationshipValidate,
```
Comment (deepu105) [inline generators/entity/templates/src/main/java/package/domain/_Entity.java]: indent is wrong

File: generators/entity/templates/src/main/java/package/domain/_Entity.java [hunk 8]
```diff
@@ -109,13 +109,15 @@
...
+        otherEntityRelationshipName = relationships[idx].otherEntityRelationshipName,
         otherEntityRelationshipNamePlural = relationships[idx].otherEntityRelationshipNamePlural,
         relationshipName = relationships[idx].relationshipName,
         relationshipFieldName = relationships[idx].relationshipFieldName,
         relationshipFieldNamePlural = relationships[idx].relationshipFieldNamePlural,
         joinTableName = entityTableName + '_'+ getTableName(relationshipName),
         relationshipType = relationships[idx].relationshipType,
+		relationshipValidateRules = relationships[idx].relationshipValidateRules,
```
Comment (deepu105) [inline generators/entity/templates/src/main/java/package/domain/_Entity.java]: indent

File: generators/entity/templates/src/main/java/package/domain/_Entity.java [hunk 9]
```diff
@@ -135,14 +137,23 @@
     _%>
     @OneToMany(mappedBy = "<%= otherEntityRelationshipName %>")
     @JsonIgnore
+
```
Comment (deepu105) [inline generators/entity/templates/src/main/java/package/domain/_Entity.java]: remove this blank line

File: generators/entity/templates/src/main/java/package/domain/_Entity.java [hunk 10]
```diff
@@ -135,14 +137,23 @@
...
+	
     <%_     if (hibernateCache != 'no') { _%>
     @Cache(usage = CacheConcurrencyStrategy.NONSTRICT_READ_WRITE)
     <%_     } _%>
     private Set<<%= otherEntityNameCapitalized %>> <%= relationshipFieldNamePlural %> = new HashSet<>();
 
     <%_ } else if (relationshipType == 'many-to-one') { _%>
     @ManyToOne
+	<%_ 
+	if (relationshipValidate == true) {
```
Comment (deepu105) [inline generators/entity/templates/src/main/java/package/domain/_Entity.java]: indent is wrong, should be 4 space

Comment (deepu105) [inline generators/entity/templates/src/main/java/package/domain/_Entity.java]: unnecessary if condition, the same is checked again in the next if

File: generators/entity/templates/src/main/webapp/app/entities/_entity-management-dialog.html [hunk 11]
```diff
@@ -210,6 +211,16 @@ <h4 class="modal-title" id="my<%= entityClass %>Label" translate="<%= keyPrefix
         </div>
             <%_ } _%>
         <%_ } _%>
+        <%_ if (relationships[idx].relationshipValidate == true) { _%>
+            <div ng-show="editForm.<%= relationshipName %>.$invalid">
```
Comment (deepu105) [inline generators/entity/templates/src/main/webapp/app/entities/_entity-management-dialog.html]: indent is wrong

**Comment by deepu105** [comment]
I guess this shouldnt be restricted to many-to-one alone, the user should be able to choose the validation for any relationship type

File: generators/entity/index.js (hunk 4 above)
Comment (dornala) [inline generators/entity/index.js]: Okk. Alright then I will do it for all.. Cool. I am loving this stuff.

**Comment by deepu105** [comment]
you need to add not null constraint on the liquibase changelog as well

**Comment by dornala** [comment]
OK Deepu. Thank you. Is spacing also so important always? Just curious. I don't do much validation on spacing.!

**Comment by deepu105** [comment]
yes its is important. Run `npm test` on your `generator-jhipster` folder to
make sure tests and lint is passing

Thanks & Regards,
Deepu

**Comment by deepu105** [comment]
you have lot of eslint errors 
```
> generator-jhipster@3.1.0 pretest /home/travis/build/jhipster/generator-jhipster
> eslint .
/home/travis/build/jhipster/generator-jhipster/generators/entity/index.js
  276:2   error  Mixed spaces and tabs                                     no-mixed-spaces-and-tabs
  276:14  error  Expected indentation of 16 space characters but found 12  indent
  573:21  error  Expected indentation of 16 space characters but found 20  indent
  574:13  error  Expected indentation of 16 space characters but found 12  indent
  575:13  error  Expected indentation of 16 space characters but found 12  indent
  576:13  error  Expected indentation of 16 space characters but found 12  indent
  577:13  error  Expected indentation of 16 space characters but found 12  indent
  578:25  error  Expected indentation of 20 space characters but found 24  indent
  579:25  error  Expected indentation of 20 space characters but found 24  indent
  580:21  error  Expected indentation of 16 space characters but found 20  indent
  581:21  error  Expected indentation of 16 space characters but found 20  indent
  582:25  error  Expected indentation of 20 space characters but found 24  indent
  583:25  error  Expected indentation of 20 space characters but found 24  indent
  584:21  error  Expected indentation of 16 space characters but found 20  indent
  585:21  error  Expected indentation of 16 space characters but found 20  indent
  586:25  error  Expected indentation of 20 space characters but found 24  indent
  587:25  error  Expected indentation of 20 space characters but found 24  indent
  588:21  error  Expected indentation of 16 space characters but found 20  indent
  589:17  error  Expected indentation of 12 space characters but found 16  indent
✖ 19 problems (19 errors, 0 warnings)
npm ERR! Test failed.  See above for more details.
The command "$JHIPSTER_INSTALL/02-install-jhipster-stack.sh" failed and exited with 1 during .
Your build has been stopped.
```

you can check it directly in your local code by running `npm run eslint` before you push to the PR

**Comment by dornala** [comment]
Wow lot of good stuff. Thanks a lot Deepu.

File: generators/entity/index.js (hunk 6 above)
Comment (dornala) [inline generators/entity/index.js]: Didn't notice it...

File: generators/entity/index.js (hunk 4 above)
Comment (dornala) [inline generators/entity/index.js]: @deepu105  I am adding required field to relationships selected as many to one , many to many and one to one. that's enough right?  many to one I don't think its required. Right?
